pkg/lib/collections/priority_queue.go

```diff
@@ -0,0 +1,205 @@
+package collections
+
+import (
+	"container/heap"
+	"errors"
+	"sync"
+)
+
+const (
+	InitialQueueCapacity = 64
+)
+
+var (
+	ErrEmptyQueue error = errors.New("queue is empty")
+	ErrNoMatch    error = errors.New("no items matched")
+)
+
+// PriorityQueue contains items of type T, and allows you to enqueue
+// and dequeue items with a specific priority. Items are dequeued in
+// highest priority first order.
+type PriorityQueue[T any] struct {
+	internalQueue queueHeap
+	mu            sync.Mutex
+}
+
+// MatchingFunction can be used when 'iterating' the priority queue to find
+// items with specific properties.
+type MatchingFunction[T any] func(possibleMatch T) bool
+
+// NewPriorityQueue creates a new ptr to a priority queue for type T.
+func NewPriorityQueue[T any]() *PriorityQueue[T] {
+	q := &PriorityQueue[T]{
+		internalQueue: make(queueHeap, 0, InitialQueueCapacity),
+	}
+	heap.Init(&q.internalQueue)
+	return q
+}
+
+// Enqueue will add the item specified by `data` to the queue with the
+// the priority given by `priority`.
+func (pq *PriorityQueue[T]) Enqueue(data T, priority int) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+
+	pq.enqueue(data, priority)
+}
+
+// enqueue is a lock-free version of Enqueue for internal use when a
+// method already has a lock.
+func (pq *PriorityQueue[T]) enqueue(data T, priority int) {
+	heap.Push(
+		&pq.internalQueue,
+		&heapItem{
+			value:    data,
+			priority: priority,
+		},
+	)
+}
+
+// Dequeue returns the next highest priority item, returning both
+// the data Enqueued previously, and the priority with which it was
+// enqueued. An err (ErrEmptyQueue) may be returned if the queue is
+// currently empty.
+func (pq *PriorityQueue[T]) Dequeue() (T, int, error) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+
+	return pq.dequeue()
+}
+
+// dequeue is a lock-free version of Dequeue() for use by internal
+// methods that already have a lock
+func (pq *PriorityQueue[T]) dequeue() (T, int, error) {
+	if pq.IsEmpty() {
+		return *new(T), 0, ErrEmptyQueue
+	}
```

Comment:
the only valid error is when dequeueing an empty queue. Since we are using generics and strong typing, we shouldn't face casting error. I hate nils, but I also hate returning tuples and returning error when there can only be a single error possible as it can complicate caller handling of dequeues. 

This is a personal preference and what you are doing here is perfectly fine

Reply:
I agree, but there's some go-isms here that make it unpleasant.  One is, I think is that T can mask that it is a pointer and there's no clean way to specify (that I've found so far) as a generic type constraint so that returning nil will always be valid. If we accept *T we end up actually with **bufTask. 

What I'd much prefer is an Option type (Optional in Java8) but I gather they're frowned upon in Go.  Will see if we can wrap it into a useful type that we can return pointers to, perhaps something like 

```go
type QueueItem[T any] struct {
    val T 
    prio int 
}
```

and then we can more easily return a ptr to a concrete type.

Reply:
Implemented in e4535c4a and looks okay actually